src/Allocator/Allocator.hs

```diff
@@ -0,0 +1,64 @@
+module Allocator.Allocator where
+
+import Env
+import Manager
+import Foundation
+import Control.Effect.Error
+import Control.Effect.State
+import Data.HashMap.Lazy
+import Control.Monad
+
+-- map of processes to its pages
+type PMap = HashMap ProcessId [Page 'Ram]
+type Ptr = Int
+newtype Size = Size { unSize :: CountOf Word8 } deriving (Eq, Show)
+
+data AllocError = OutOfMemory deriving (Eq, Ord)
+
+type AllocSig sig m = ( Has (State PMap) sig m
+                      , Has (Throw AllocError) sig m
+                      , Has (Catch AllocError) sig m)
+
+
+
+-- TODO mock of env, actually should be exported from somewhere
+env :: Env
+env = Env 1024 0 0
+
+class Allocator a where
+    alloc   :: (ManagerSig sig m, Has (State PMap) sig m) => a -> Size -> m (Maybe Ptr)
+    realloc :: (ManagerSig sig m, Has (State PMap) sig m) => a -> Ptr -> Size -> m (Maybe Ptr)
+    -- TODO free should return an error or (Nothing, ok, anything else?), is Maybe sufficient for that?
```

Comment:
Since we're mimicking C malloc interface, we should do it like there(so `Maybe` is sufficient). On the other hand, we're creating model in haskell, so we can make it "haskellysh". But I'm more inclined to mimicking C interface, since this is only a lab...